File: ProjectEuler/001-050/S27.cs
using System.Collections.Generic;

namespace ProjectEuler
{
    public class S27 : ISolution
    {
        struct ABAndMaxPrime
        {
            internal int A;
            internal int B;
            internal int PrimeCount;
        }

        public string GetAnswer()
        {
            var primes = new List<ABAndMaxPrime>();

            for (int a = -999; a < 1000; a++)
            {
                for (int b = -999; b < 1000; b++)
                {
                    var tmp = new ABAndMaxPrime
                    {
                        A = a,
                        B = b,
                        PrimeCount = 0
                    };

                    for (int n = 0; ; n++)
                    {
                        int formula = n * n + a * n + b;
                        if (Utils.IsPrime(formula))
                        {
                            tmp.PrimeCount++;
                        }
                        else
                        {
                            primes.Add(tmp);
                            break;
                        }
                    }
                }
            }

            var max = new ABAndMaxPrime();
            foreach (var prime in primes)
            {
                if (prime.PrimeCount > max.PrimeCount)
                {
                    max = prime;
                }
            }

            return (max.A * max.B).ToString();
        }
    }
}

File: src/001-050/S27.cc
#include "S27.h"
#include "../Utils.h"

using namespace std;

namespace ProjectEuler
{
std::string S27::GetAnswer()
{
	std::vector<S27::ABAndMaxPrime> primes;

	for (int a = -999; a < 1000; a++)
	{
		for (int b = -1000; b <= 1000; b++)
		{
			S27::ABAndMaxPrime tmp;
			tmp.A = a;
			tmp.B = b;
			tmp.PrimeCount = 0;

			for (int n = 0;; n++)
			{
				int formula = n * n + a * n + b;
				if (Utils::IsPrime(std::abs(formula)))
				{ // C# implementation might imply checking primality of result
					tmp.PrimeCount++;
				}
				else
				{
					primes.push_back(tmp);
					break;
				}
			}
		}
	}

	S27::ABAndMaxPrime max = { 0, 0, -1 };
	for (const auto &prime : primes)
	{
		if (prime.PrimeCount > max.PrimeCount)
		{
			max = prime;
		}
	}

	return std::to_string(max.A * max.B);
}

} // namespace ProjectEuler
